react/week4/mainForm.js

```diff
@@ -0,0 +1,28 @@
+import React, { useContext } from "react";
+import { StateContext } from "./context.js";
+
+export default function MainForm() {
+  const { users, loading, fetchUsers } = useContext(StateContext);
```

Comment:
Remember Single responsibility principle. This component has more than one responisbility. The display and input should be in different components

Reply:
Separated them, is it how it should be?

Reply:
Or something along the lines this maybe:

```
import React, { useContext, createContext, useState, useEffect } from "react";

function App() {
  return (
    <UsersSearch/>
  )
}

export const UserSearchContext = createContext();

function UsersSearch() {
  const [isLoading, setIsLoading] = useState(false);
  const [users, setUsers] = useState([]);
  const [userInput, setUserInput] = useState("");

  const handleErrors = (response) => {
    if (!response.ok) {
      throw Error(response.statusText);
    }
    return response;
  };

  useEffect(() => {
    (async () => {
      try {
        setIsLoading(true);
        const response = await fetch(
          `https://api.github.com/search/users?q=${userInput}`
        );
        handleErrors(response);
        const data = await response.json();
        const usersData = await data.items;
        setUsers(usersData);
      } catch (err) {
        console.log(err);
      } finally {
        setIsLoading(false);
      }
    })();
  }, [userInput])

 
  return (
    <UserSearchContext.Provider value={{ users, userInput, setUserInput }}>
      <SearchForm />
      { isLoading ? <Loading /> : <UsersDisplay /> }
    </UserSearchContext.Provider>
  );
}

function Loading() {
  return (
    <div>Loading... </div>
  )
}

function UsersDisplay() {
    const { users} = useContext(UserSearchContext);
  
    return (
      <div className="App">
        {users.length === 0 || !users ? (
          <h2>No results</h2>
        ) : (
          users.map((user) => <div key={user.id}> {user.login}</div>)
        )}
      </div>
    );
  }

function SearchForm() {
    const { userInput, setUserInput } = useContext(UserSearchContext);
  
    return (
      <div className="App">
        <h1>Github user searcher</h1>
        <div>
          <input
            placeholder="Search for user"
            value={userInput}
            onKeyUp={(e) => {
              setUserInput(e.target.value);
            }}
          />
        </div>
      </div>
    );
  }

export default App;
```